### PR TITLE
Implement DomainMetadata Action

### PR DESCRIPTION
**Pull request by dabao521**
## Description
This change implements [DomainMetadata](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#domain-metadata) Action in the delta spec.

## How was this patch tested?
New unit tests are added.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala [hunk 1]
```diff
@@ -297,6 +301,57 @@ private[delta] class ConflictChecker(
...
+   */
+  private def checkIfDomainMetadataConflict(): Unit = {
+    if (!DomainMetadataUtils.domainMetadataSupported(currentTransactionInfo.protocol)) {
+      return
+    }
+    // Note [[currentTransactionInfo.domainMetadatas]] can't have duplicate domains otherwise
+    // it would have failed [[validateDuplicateDomainMetadata]] at beginning of the transaction's
+    // validation.
+    val currentDomainMetadatas =
+    currentTransactionInfo.domainMetadatas.map(action => action.domain -> action)
```
Comment (imback82) [inline core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:320]:
```suggestion
      currentTransactionInfo.domainMetadatas.map(action => action.domain -> action)
```

**Review by imback82 — APPROVED** [review]
few nits, but lgtm

**Review by imback82 — APPROVED** [review]
few nits, but lgtm

**Review by ryan-johnson-databricks — COMMENTED** [review]
> Description
> 
> This change implements [DomainMetadata](https://github.com/dabao521/delta/pull/new/domainMetadata) Action in the delta spec.

I don't think that's the right link. Can you link the actual merged PR instead?

**Comment by ryan-johnson-databricks** [comment]
> > Description
> > This change implements [DomainMetadata](https://github.com/dabao521/delta/pull/new/domainMetadata) Action in the delta spec.
> 
> I don't think that's the right link. Can you link the actual merged PR instead?

... or just link to the actual public spec: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#domain-metadata

**Review by dabao521 — COMMENTED** [review]
@ryan-johnson-databricks  thanks for the review . I have addressed your comments and fixed the spec link in the description. Let me know if it looks good to you.

File: core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala [hunk 2]
```diff
@@ -329,28 +324,30 @@ private[delta] class ConflictChecker(
...
+    val mergedDomainMetadata = mutable.Buffer.empty[DomainMetadata]
     // Update the [[DomainMetadata]] to the merged one.
-    val updatedActions = currentTransactionInfo.actions.map {
+    val updatedActions: Seq[Action] = currentTransactionInfo.actions.map {
       case domainMetadata: DomainMetadata =>
-        // Note the find will always succeed, because [[mergedDomainMetadatas]] already contains
-        // every domain from the transaction.
-        mergedDomainMetadatas.find(_.domain == domainMetadata.domain).getOrElse(domainMetadata)
+        val mergedAction = resolveConflict(domainMetadata)
+        mergedDomainMetadata += resolveConflict(domainMetadata)
```
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:344]: I don't think we want to call `resolveConflict` twice?
```suggestion
        mergedDomainMetadata += mergedAction
```

Comment (dabao521) [reply core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala]: Whoops, just fixed

File: core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala [hunk 3]
```diff
@@ -301,6 +303,54 @@ private[delta] class ConflictChecker(
...
+        case (domain, None) => domain
+        case (_, Some(_)) =>
+          // Any conflict not specifically handled by a previous case must fail the transaction.
+          throw new io.delta.exceptions.ConcurrentTransactionException(
+            s"A conflicting metadata domain ${domainMetadataFromCurrentTransaction.domain} is " +
+              "added.")
+      }
+
+    val mergedDomainMetadata = mutable.Buffer.empty[DomainMetadata]
+    // Update the [[DomainMetadata]] to the merged one.
```
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:340]:
```suggestion
    // Resolve physical [[DomainMetadata]] conflicts (fail on logical conflict).
    val mergedDomainMetadata = mutable.Buffer.empty[DomainMetadata]
```

File: core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 4]
```diff
@@ -1128,6 +1136,7 @@ trait OptimisticTransactionImpl extends TransactionalWrite
             assertMetadata(m)
           case p: Protocol =>
             recordProtocolChanges(snapshot.protocol, p, isCreatingNewTable)
+          case _: DomainMetadata => numOfDomainMetadatas += 1
```
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1139]: nit: all the other cases in this match have a newline after the `=>`;
should we match style here for consistency?

**Review by ryan-johnson-databricks — APPROVED** [review]
LGTM other than a couple of nits

File: core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala (hunk 3 above)
Comment (dabao521) [reply core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala]: fixed

File: core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 4 above)
Comment (dabao521) [reply core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala]: fixed

**Pull request closed without merging**